normandy/base/middleware.py

```diff
@@ -117,3 +122,150 @@ def process_request(self, request):
                 }
             )
         return response
+
+
+class OIDCEndpointRequestError(Exception):
+    """Happens when the server-to-server communication with the OIDC endpoint
+    succeeds but the OIDC endpoints responds with a status code > 200 and less
+    than 500."""
+
+
+class OIDCAccessTokenAuthorizationMiddleware:
+    """Middleware that looks for 'Authorization: Bearer access_token' headers in
+    the request and uses this access token to use an OIDC provider to turn that into
+    a user profile. If that works, the request becomes signed in as this user.
+
+    Note, it only signs in the user's request. Not session. Since this is not
+    an authentication backend, there is no 'Set-Cookie' header returned in the
+    response.
+
+    It's admittedly confusing to mix the terms authentication and authorization.
+    Arguably the client has already *authenticated* and basically converted her
+    password for an access token. Here in the context of Normandy we kill two
+    birds with one stone: check that access token and use the user profile that
+    comes out to sign the user into the rest of the Django request.
+
+    The flow is as follows:
+
+    1. Client trades email/username and password with the OIDC for an access token.
+    2. The access token is sent to us here.
+    3. We send it to the OIDC provider in return for a user profile
+    4. We extract the email from the user profile (and first- and last name)
+    5. Turn the email into a Django User model instance.
+    6. If possible, extracts the first- and last name and updates the User instance persistently.
+    7. Sets that user as the user of the request thus enabling `request.user.is_active`.
+
+    NOTE! When the bearer token is present, it does a network request to the OIDC provider
+    every single time. There is no caching. Even for identical access tokens used
+    frequently. The justification for this at the Normandy Admin UI requests are
+    not high-traffic to performance critical. Also, we don't have a cache framework
+    set up.
```

Comment:
> Also we don't have a cache frame work set up

IIRC, Django defaults to an in-memory cache, which would still be beneficial.

Reply:
I'm just not convinced it's worth it. But perhaps you have a better feeling for how the Admin UI works. I don't expect there to be that much traffic anyway. I'd rather stall and incrementally add this later if we find that it's slowing 
things down. 

We'd need to try to make sure the TTL on that caching matches the TTL on the access token itself. I think it's set to 2h according to the idToken I get in delivery console. 

Mind you, the in-memory cache is per Python process and if we have many workers per EC2 there's little cache hits.

Reply:
I'm not arguing for or against storing things in the cache. I'd be happy to leave the code as it is. I was just objecting to that last sentence, as it isn't strictly true.

Reply:
Sentence removed.